testtools/UART_interface/serial_connect.py

```diff
@@ -0,0 +1,160 @@
+        while output < 4:
+            time.sleep(1)
+            output = ser.in_waiting
+            print("%d bytes in waiting" %output)
+
+    if ser.readable():
+        output = ser.readline(ser.in_waiting)
+        print(output.decode(encoding='utf-8'))
+        try:
+            # File must exist to write to it
+            file.writelines(output.decode(encoding='utf-8'))
+        except:
+            pass
+
+        return output
+    else:
+        try:
+            time.sleep(2)
+            ser.open()
+            serial_read(ser, message, file, first_read=True)
+        except:
+            return False
+
+
+def write_read(ser, input_file, output_file):
+    if input_file:
+        # set wait between read/write
+        wait = (serial_settings.bits_to_cache/serial_settings.baud_rate)
+        with open(input_file) as fp:
+            # initialize input/output_file
+            line = fp.readline()
+            f = open(output_file, 'w+') # Output file
+            while line:
+                time.sleep(.1)
+                print("Sending %s" %line)
+
+                # Attempt to write to serial port
+                if not serial_write(ser, line, f):
+                    print("burger king")
+                    f.close()
+                    break
+                time.sleep(wait)
+
+                # Attempt to read serial port
+                output = serial_read(ser, line, f, first_read=True)
+
+                while(output):
+                    time.sleep(wait)
+                    output = serial_read(ser, line, f)
+                line = fp.readline()
+
+            f.close()
+
+def run():
+    if 'posix' in os.name:
+        # enable read write access on port
+        # os.chmod(serial_settings.port, 666)
+
+        # check to see that mxchip firmware drop path resolves
+        for x in range(serial_settings.wait_for_flash):
+            exists = os.path.exists('/media/newt/AZ31661')
+            if exists:
+                break
+            time.sleep(1)
+        if x >= (serial_settings.wait_for_flash-1):
+            print("MXCHIP drop path did not resolve, chip did not flash successfully.")
+            raise FileExistsError
+
+    print(serial_settings.baud_rate)
+    ser = serial.Serial(port=serial_settings.port, baudrate=serial_settings.baud_rate)
+
+    # print(ser.writable())
+    # print(ser.readable())
+    time.sleep(.5)
+    # Print initial message
+    output = ser.readline(ser.in_waiting)
+    print(output.strip().decode(encoding='utf-8'))
+    while(output):
+        output = ser.readline(ser.in_waiting)
+        print(output.strip().decode(encoding='utf-8'))
+
+    write_read(ser, serial_settings.input_file, serial_settings.output_file)
+
+    while(ser.in_waiting):
+        time.sleep(.2)
+        output = ser.readline(ser.in_waiting)
+        print(output.strip().decode(encoding='utf-8'))
+        # output = input("howzit")
+        # serial_write(ser, output)
+    ser.close()
```

Comment:
>close [](start = 8, length = 5)

I am seeing examples of this library showing reset_output_buffer() prior to close(). Maybe we should consider adding that for cleaner shutdown?

Reply:
flush() also works. I'm cool with that.